src/Storages/StorageAggregatingMemory.cpp

```diff
@@ -0,0 +1,513 @@
+
+            auto converted_field = convertFieldToType(literal->value, *column_type);
+            if (!converted_field.isNull() && data.raw_key_columns[key_position]->empty())
+            {
+                data.raw_key_columns[key_position]->insert(converted_field);
+            }
+        }
+    }
+};
+
+using AggregationKeyVisitor = AggregationKeyMatcher::Visitor;
+
+static std::optional<AggregationKeyVisitor::Data> getFilterKeys(const Aggregator::Params & params, const SelectQueryInfo & query_info)
+{
+    if (params.keys.empty())
+        return {};
+
+    const auto & select = query_info.query->as<ASTSelectQuery &>();
+    ASTPtr where = select.where();
+    if (!where)
+        return {};
+
+    AggregationKeyVisitor::Data data(params);
+    AggregationKeyVisitor(data).visit(where);
+
+    if (data.failed || !data.hasAllKeys())
+        return {};
+
+    return data;
+}
+
+/// AggregatingOutputStream is used to feed data into Aggregator.
+class AggregatingOutputStream : public IBlockOutputStream
+{
+public:
+    AggregatingOutputStream(StorageAggregatingMemory & storage_, const StorageMetadataPtr & metadata_snapshot_, ContextPtr context_)
+        : storage(storage_),
+          metadata_snapshot(metadata_snapshot_),
+          context(context_),
+          variants(*(storage.many_data)->variants[0]),
+          key_columns(storage.aggregator_transform->params.keys_size),
+          aggregate_columns(storage.aggregator_transform->params.aggregates_size)
+    {
+        expression_actions = std::make_shared<ExpressionActions>(storage.analysis_result.before_aggregation);
+    }
+
+    // OutputStream structure is same as source (before aggregation).
+    Block getHeader() const override { return storage.src_block_header; }
+
+    void write(const Block & block) override
+    {
+        storage.src_metadata_snapshot->check(block, true);
+
+        StoragePtr source_storage = storage.source_storage;
+        auto query = metadata_snapshot->getSelectQuery();
+
+        StoragePtr block_storage
+            = StorageValues::create(source_storage->getStorageID(), source_storage->getInMemoryMetadataPtr()->getColumns(), block, source_storage->getVirtuals());
+
+        ContextMutablePtr local_context = Context::createCopy(context);
+        local_context->addViewSource(block_storage);
+
+        InterpreterSelectQuery select(query.inner_query, local_context, SelectQueryOptions(QueryProcessingStage::WithMergeableState));
+        auto select_result = select.execute();
+
+        BlockInputStreamPtr in;
+        in = std::make_shared<MaterializingBlockInputStream>(select_result.getInputStream());
+        in = std::make_shared<SquashingBlockInputStream>(
+            in, context->getSettingsRef().min_insert_block_size_rows, context->getSettingsRef().min_insert_block_size_bytes);
+
+        in->readPrefix();
+
+        bool no_more_keys = false;
+        while (Block result_block = in->read())
```

Comment:
we have one block passed in, can we just `aggregator.mergeBlock(block)` directly instead of going through the whole in-memory pipe ?